### DISCUSSION TITLE
How to remove the Layout on 404 page

### DISCUSSION THREAD

**Discussion by MGMehdi** [vercel/next.js#37311] [Help] [answered]
Hello,
I can set a Layout but when I go to a wrong path it keep showing on 404 error page how can I change this ?
Here is my code

**_app.tsx**
```js
import Layout from "../components/Layout";

import type { AppProps } from "next/app";

export default function MyApp({ Component, pageProps }: AppProps) {
  return (
    <>
      {/* @ts-ignore */}
      <Layout>
        <Component {...pageProps} />
      </Layout>
    </>
  );
}
```
**Layout.tsx**
```js
import React from "react";
import type { NextPage } from "next";

type HomeProps = {
  window: any;
  children: any;
};

const Layout: NextPage<HomeProps> = ({ window, children }: HomeProps) => {
  return (
    <>
      <h1>This is not an AppBar</h1>
      <main>{children}</main>
    </>
  );
};

export default Layout;
```
Thank you

**Comment by icyJoseph** [comment, marked as answer]
So you don't want to show the layout on 404 pages?  (that's a bit unexpected, but doable)

The issue happens in `_app`, because this component wraps all pages. We need to skip doing this for 404 pages.

There's two ways, as far as I can tell:

```tsx
import type { AppProps } from "next/app";
import { useRouter } from "next/router";

function MyApp({ Component, pageProps }: AppProps) {
  const router = useRouter();
  console.log(router);

  if(router.pathname === '_error') return <Component {...pageProps} />;

  return (
      <Layout>
        <Component {...pageProps} />
      </Layout>
   );
}

export default MyApp;
```

Or the most, correct one, to use per-page layouts. Read more here, https://nextjs.org/docs/basic-features/layouts#per-page-layouts

```tsx
// _app
import type { ReactElement, ReactNode } from 'react'
import type { NextPage } from 'next'
import type { AppProps } from 'next/app'

type NextPageWithLayout = NextPage & {
  getLayout?: (page: ReactElement) => ReactNode
}

type AppPropsWithLayout = AppProps & {
  Component: NextPageWithLayout
}

function commonLayout(page) {
  return (
    <Layout>
      <NestedLayout>{page}</NestedLayout>
    </Layout>
  )
}

export default function MyApp({ Component, pageProps }) {
  // Use the layout defined at the page level, if available
  const getLayout = Component.getLayout || commonLayout

  return getLayout(<Component {...pageProps} />)
}
```

Define, `_error.tsx` under pages, or if you just want to target `404`, create `404.tsx` 

```tsx
// _error.tsx
import type { ReactElement } from 'react'
import type { NextPageContext } from "next";
import Error, { ErrorProps } from "next/error";

function CustomError({ statusCode }: ErrorProps) {
  return <Error statusCode={statusCode} />;
}

CustomError.getInitialProps = ({ res, err }: NextPageContext) => {
  const statusCode = res ? res.statusCode : err ? err.statusCode : 404;
  return { statusCode };
};

CustomError.getLayout = function getLayout(page: ReactElement) {
  return <>{page}</>;
};

export default CustomError;
```

for 404's only:

```tsx
// 404.tsx
import type { ReactElement } from "react";
import Error from "next/error";

function Custom404() {
  return <Error statusCode={404} />;
}

Custom404.getLayout = function getLayout(page: ReactElement) {
  return <>{page}</>;
};

export default Custom404;
```

This last one, only 404, I think, it's best, because it can be pre-rendered, and it can live side by side with `_error.tsx` if I remember correctly, (you can have both), read more here: https://nextjs.org/docs/advanced-features/custom-error-page#404-page

**Reply by MGMehdi** [reply]
Thanks I didn't see this page https://nextjs.org/docs/advanced-features/custom-error-page#404-page

**Reply by sneko** [reply]
@icyJoseph I'm using appDir and using old `pages` directory to catch 404 errors. It works but I get an hydratation error that breaks some JS bahviors.
```
Warning: validateDOMNesting(...): <head> cannot appear as a child of <div>.
```

Here is the basic error page:
https://github.com/inclusion-numerique/mediature/blob/dev/apps/main/src/pages/404.tsx

I tried to:
* simplify at the maximum (no reuse of a layout, just a div with text)
* use the getLayout
* add a `_app.tsx` to set the `html+body` structure (even this case complains about `html` behind a child of a `div`... which makes no sense since nothing should be above this)

but in all cases it seems something is set wrapping what I define at the higher level. Are you aware of something that could produce this weird behavior?

Thank you,

---

For example if I set in my `_app.tsx`:
```tsx
export default function MyApp({ Component, pageProps }) {
  return (
    <html lang="fr">
      <body>
        <div>Salut</div>
      </body>
    </html>
  );
}
```

I will see in the raw HTML from the server:
```html
<!DOCTYPE html><html lang="fr"><head>...</head><body><html lang="fr"><div id="__next"><body><div>Salut</div></div></body></html><script src="[/_next/static/chunks/react-refresh.js?ts=1678815366010](http://localhost:3000/_next/static/chunks/react-refresh.js?ts=1678815366010)"></script><script id="__NEXT_DATA__" type="application/json">{"props":{"pageProps":{}},"page":"/404","query":{},"buildId":"development","runtimeConfig":{"appMode":"test","appVersion":"v0.0.0-20230314160934-2cffe3245a45"},"nextExport":true,"autoExport":true,"isFallback":false,"locale":"fr","locales":["fr"],"defaultLocale":"fr","scriptLoader":[]}</script></body></html>
```

We can see here we have 2 `html`. That's why my `404.tsx` does not work with its layout. I would be okey to remove `html+body` to use my components but I need them to set some JS scripts tags :( 

(I need this since `appDir` is not ready at all for managing errors properly)

**Reply by icyJoseph** [reply]
The way the pages directory works is fundamentally different than appDir.

Achieving inter-op between them sounds like a bad idea.

The pages directory uses `_document`, even if not defined, to set the foundation of pages. This injects html, head, and body tags, as well as the `next script tag`.

The `_app` page is the first component rendered by the `next script tag` entry point, and it is rendered within a div with `id` set to `__next`.

It is as this depth, `html (head, body (div[id="__next"]))` that all pages are mounted, including error pages. This is all out of your control.

In your example, the root, as per appDir requirement, contains all of the top level tags, html, head, body, and that is being wrapped by `__next`

I'd not explore further, this seems like a dead-end, and at best full with pitfalls.

You'd need to basically, implement your `_document`, `_app`, and `404` pages using the page directory rules, and preferably with dedicated components, aka, not using any of the special constructs of appDir.

**Reply by arturamorfati** [reply]
ty so much!

**Reply by reza-ebrahimi** [reply]
@icyJoseph I have same problem with app router, how to show `notFound` page without layout?

**Reply by shriekdj** [reply]
@icyJoseph I also have the same issue with nextjs 13 app router notFound Page. i want it to without layout

**Reply by icyJoseph** [reply]
It looks like using route groups for not-found is not yet possible:

There's a workaround here, using a `[...notFound]` https://github.com/vercel/next.js/discussions/50034

**Reply by Y-k-Y** [reply]
@reza-ebrahimi @shriekdj Did you guys find solution for app router?

**Reply by reza-ebrahimi** [reply]
@Y-k-Y No :(

**Comment by ezeparziale** [comment]
Any solution for app router? I tried the workaround from https://github.com/vercel/next.js/discussions/50034 but doesnt work

**Reply by autoool** [reply]
haha,me too

**Reply by eliac7** [reply]
Well, I am in the same boat; has anyone figured it out?

**Reply by revengemiroz** [reply]
one thing i have found is dont include you nav bar and footer in the main app layout instead you need to wrap the nav and footer in each page layout.js. That way you 404 will cover the entire page and not show the layout

**Reply by shriekdj** [reply]
i also found the solution https://github.com/vercel/next.js/discussions/37311#discussioncomment-6993373

**Comment by ezeparziale** [comment]
I am using this folder organization.

![Captura de pantalla 2023-09-13 a la(s) 09 19 24](https://github.com/vercel/next.js/assets/82482391/cc544405-35f7-48a1-81e5-ed6a10da71b3)

In this case I have Nav bar in root page and dashboard, but not in login page or auth section. In 404 error I got a page without any layout. You need to define the `not-found.tsx`

https://nextjs.org/docs/app/building-your-application/routing/pages-and-layouts
https://nextjs.org/docs/app/api-reference/file-conventions/not-found

**Reply by shriekdj** [reply]
i also done something similar to this

**Comment by shriekdj** [comment]
Hello Everyone, I Got The Solution We Just Need to Restructure Our Code.
Before Restructuring My NextJS Code Look Like this.

```
src/
    app/
        [...not_found]/page.tsx
        about/page.tsx
        contact/page.tsx
        layout.tsx
        not-found.tsx
   components/
   utils/
```

After Restructuring The Folders and Files Look Like this.

```
src/
    app/
       (default_site)/
            [...not_found]/page.tsx
            about/page.tsx
            contact/page.tsx
            layout.tsx
       (error_layout)/
            layout.tsx
            not-found.tsx
   components/
   utils/
```

Here are 2 layout.tsx Both Containing Root Layout For Example layout.tsx in (default_site)/ contains the global AdSense Code with Google Analytics.

But Error Page Like 404 Page only Contains the google analytics and there is no adsense code there at all.

As Per NextJS Docs We Can Have Multiple RootLayout for Multiple Parts of our site like sub applications by this way.

**Reply by shriekdj** [reply]
@Y-k-Y @reza-ebrahimi try this solution

**Reply by Y-k-Y** [reply]
@shriekdj Yes, recently i figured out the route group and did similar work as you did. But is it possible to nest error pages in a route group? I've tried but it seems Next couldn't find out error pages in route group. The default error page is showing instead

**Reply by benjipeto** [reply]
I tried this solution, however came across the same problem as @Y-k-Y, is there anything else you did differently?

**Reply by benjipeto** [reply]
In the end, this did help! 
However for the first time, trying to completely follow your solution I got some error _(sorry, don't remember exactly what kind)_. Just tried again, because reading the documentation this really makes sense, and it's a quite nice solution. 

For me, I organised everything in a folder with parentheses including the rootLayout, and left the `not-found.tsx `page in the /app directory, with its own layout. So basically did the same @shriekdj, except moving the `not-found.tsx `and its layout in their separate folder. _(But thinking of now, I must have been doing something wrong that it didn't work to be fair 😄 )_

Thanks again 🙏🏽 you are a legend 💯

**Reply by shriekdj** [reply]
happy to help @benjipeto

**Reply by benjipeto** [reply]
Thank you @shriekdj ! For the time being it works like a charm, the only thing I different than you is that I did not create a folder for the `not-found.tsx` I left it in the root with its 'special' layout. However according to the [docs](https://nextjs.org/docs/app/building-your-application/routing/route-groups), this is the best and possibly only way. Thanks once again for your help! Have a good one ;)

**Reply by rimoslav** [reply]
Thanks @shriekdj, after months of digging this solution helped. I'll spread the word on other open issues and StackOverflow questions :) 
If I understood @benjipeto correctly I think I did the same thing as he did.
I've split `app/layout.tsx` into two parts. The first part that I want to wrap the whole app i've placed in `app/layout.tsx` and added `app/not-found.tsx` (for custom design, not really necessary from the perspective of this issue).
The rest of the logic and wrapper components that I want only on non-404 routes I've placed into `app/(default_site)/layout.tsx`.
Now when NextJS renders the app it catches that it's a 404 route when it gets to `app/layout.tsx` and doesn't proceed to render `app/(default_site)/layout.tsx` and the rest of the app. That way I didn't need `app/(error_layout)` directory at all.

Thanks again :)

**Reply by shriekdj** [reply]
Happy to help, @rimoslav 😊

**Reply by eviannaive** [reply]
@shriekdj The problem has comfuse me some time. it's helpful~thank a lot🥹

**Reply by sauravhathi** [reply]
If the current setup is not working, try moving `[...not_found]/page.tsx` from the `(default_site)` folder to the` (error_layout)` folder.

This ensures that undefined routes use the appropriate lightweight error layout without loading unnecessary components like ads or the `sidebar/footer`.

**Comment by mfaeezshabbir** [comment]
I've found two effective solutions for this problem `using page router`:

**Method 1: Conditional Rendering in _app.js**

This approach uses the `useRouter` hook to check if the user is on the 404 page (/_error). If not, it wraps the page component with your layout. Here's the code:

```
import { useRouter } from 'next/router';

// Import your layout component
import Layout from '../components/Layout';

function MyApp({ Component, pageProps }) {
  const router = useRouter();

  return (
    <div>
      {router.pathname !== '/_error' && <Layout><Component {...pageProps} /></Layout>}
      {router.pathname === '/_error' && <Component {...pageProps} />}
    </div>
  );
}
export default MyApp;
```

> What's happening?
> 1. We import the useRouter hook and your layout component.
> 2. Inside _app.js, we use router.pathname to check if the user is on the 404 page.
> 3. If it's not the 404 page, we render the page component wrapped in your layout.
> 4. If it's the 404 page, we just render the page component without the layout.

**Method 2: Per-Page Layouts**

You  can learn more here: [https://nextjs.org/docs/basic-features/layouts#per-page-layouts](https://nextjs.org/docs/basic-features/layouts#per-page-layouts).

**Reply by dragoshuniq** [reply]
for that you need "use client" Karl, you just killed all the features of the NextJS

**Comment by ragini743** [comment]


**Reply by shriekdj** [reply]
does it give 404 response code also ?
